### PR TITLE
 Move `from_preset` to base tokenizer classes

### LINKED ISSUE #648: Move `from_preset` to base tokenizer classes
This PR would be an replication of #621 for our three **tokenizer** layers: `WordPieceTokenizer`, `BytePairTokenizer`, and `SentencePieceTokenizer`. Unlike #621 we do not need to create new classes but merely move some properties and methods to the existing ones.

Side note: Originally I wanted to move `from_preset` to `tokenizer.Tokenizer`, but preset loading differs for the three subclasses above.

Example steps for `WordPieceTokenizer`:
* Copy over following methods/properties to base class from `BertTokenizer`:
    * `presets` (return `{}`)
    * `from_preset` with string formatting params to adapt to preset and class names
* Remove `from_preset` from each subclass and fill in the string params for `from_preset` docstring.

Following #621, add a mention of `from_preset` loading to the generic constructor docstring.

This process should be repeated for every `XXTokenizer` in the `models/` folder.

### PR DESCRIPTION
**Pull request by shivance**
Closes #648

### CODE REVIEW

**Comment by shivance** [comment]
@jbischof Please review.

File: keras_nlp/tokenizers/byte_pair_tokenizer.py [hunk 1]
```diff
@@ -536,3 +537,57 @@ def _bpe_merge_and_update_cache(self, tokens):
             tokenized_words, axis=1, separator=" "
         )
         self.cache.insert(tokens, tokenized_words)
+
```
Comment (mattdangerw) [inline keras_nlp/tokenizers/byte_pair_tokenizer.py:540]: each of these classes should probably have a preset property defined now, that is empty. E.g.

https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/models/backbone.py#L33-L35

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Thanks you!!

The actual runnable code changes looks good, just one minor comment.

We will likely need to do something fancier for docstrings though. Will think through this a bit and post more here.

File: keras_nlp/tokenizers/byte_pair_tokenizer.py [hunk 2]
```diff
@@ -536,3 +537,57 @@ def _bpe_merge_and_update_cache(self, tokens):
...
+        tokenizer = keras_nlp.models.GPT2Tokenizer.from_preset(
+            "gpt2_base_en",
+        )
+        # Tokenize some input.
+        tokenizer("The quick brown fox tripped.")
+        # Detokenize some input.
+        tokenizer.detokenize([5, 6, 7, 8, 9])
+        ```
+        """
+
```
Comment (jbischof) [inline keras_nlp/tokenizers/byte_pair_tokenizer.py:564]: After adding the `preset` property, check if empty as we do for backbone ([link](https://github.com/keras-team/keras-nlp/blob/99edccfe6388a2ea7f34f57fecb32c32cada33bc/keras_nlp/models/backbone.py#L66)). Same for the other two tokenizers.

File: keras_nlp/tokenizers/byte_pair_tokenizer.py [hunk 3]
```diff
@@ -536,3 +537,57 @@ def _bpe_merge_and_update_cache(self, tokens):
...
         )
         self.cache.insert(tokens, tokenized_words)
+
+    @classmethod
+    def from_preset(
+        cls,
+        preset,
+        **kwargs,
+    ):
+        """Instantiate a GPT-2 tokenizer from preset vocabulary and merge rules.
```
Comment (jbischof) [inline keras_nlp/tokenizers/byte_pair_tokenizer.py:547]: Please remove GPT-2 language. This should be a generic docstring for BPE. Same for the other two

**Review by jbischof — CHANGES_REQUESTED** [review]
Thanks mostly documentation changes!

File: keras_nlp/models/bert/bert_tokenizer.py [hunk 4]
```diff
@@ -113,51 +111,9 @@ def __init__(
     def presets(cls):
         return copy.deepcopy({**backbone_presets, **classifier_presets})
 
-    @classmethod
-    @format_docstring(names=PRESET_NAMES)
     def from_preset(
```
Comment (mattdangerw) [inline keras_nlp/models/bert/bert_tokenizer.py:114]: After following the changes below re docstrings and `__init__subclass__` you should be able to remove the `from_preset` method here and elsewhere entirely!

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Left some comments re how we can handle the docstrings here.

Also, we are a bit of a moving target here (lots of changes this week!), but you can also mirror these changes for the albert and f_net models. Thank you!

**Comment by mattdangerw** [comment]
Looks like you also have some formatting issues on this PR, checkout the `Tests / Check the code format (pull_request)` above!

File: keras_nlp/models/bert/bert_tokenizer.py (hunk 4 above)
Comment (jbischof) [reply keras_nlp/models/bert/bert_tokenizer.py:114]: @shivance, we don't need `from_preset` in subclasses anymore! See, for example, [BertPreprocessor](https://github.com/keras-team/keras-nlp/blob/fb1eeb9054a2a13be4853a6e635e9a72d54d04db/keras_nlp/models/bert/bert_preprocessor.py#L195)

File: keras_nlp/models/xlm_roberta/xlm_roberta_tokenizer.py [hunk 5]
```diff
@@ -170,44 +169,4 @@ def from_preset(
...
-
-        config = metadata["preprocessor_config"]
-        config.update(
-            {
-                "proto": spm_proto,
-            },
-        )
-
-        return cls.from_config({**config, **kwargs})
+        return super().from_preset(cls, preset, **kwargs)
```
Comment (jbischof) [inline keras_nlp/models/xlm_roberta/xlm_roberta_tokenizer.py:172]: You need a newline at the end of each file. Are you still having issues with the `format.sh` script?

**Review by jbischof — COMMENTED** [review]
Good progress! Let us know if you get stuck. You need to run the `format.sh` script before every commit, so let us know if you're having trouble there.

File: keras_nlp/models/xlm_roberta/xlm_roberta_tokenizer.py (hunk 5 above)
Comment (shivance) [reply keras_nlp/models/xlm_roberta/xlm_roberta_tokenizer.py:172]: Hi @jbischof , I'm still addressing the comments. So work is pending.
I tend to run formatting scripts upon finishing changes for every round of review.

Should I continue with this or run it every time before commit?

Thanks.

Comment (jbischof) [reply keras_nlp/models/xlm_roberta/xlm_roberta_tokenizer.py:172]: Sorry, didn't understand this was still WIP! Up to you on how you organize your commits 😄

**Comment by shivance** [comment]
@jbischof It's ready for review now 😄

File: keras_nlp/tokenizers/sentence_piece_tokenizer.py [hunk 6]
```diff
@@ -206,3 +209,83 @@ def tokenize(self, inputs):
...
+    @classmethod
+    def from_preset(
+        cls,
+        preset,
+        **kwargs,
+    ):
+        """Instantiate {{model_name}} tokenizer from preset vocabulary.
+
+        Args:
+            preset: string. Must be one of {{preset_names}}.
```
Comment (mattdangerw) [inline keras_nlp/tokenizers/sentence_piece_tokenizer.py:226]: Surround with quotes.

**Review by mattdangerw — APPROVED** [review]
LGTM! Thank you!

Just found a few small nits that need fixing.

**Review by jbischof — APPROVED** [review]
Looks good in general, but please follow @mattdangerw's suggestions and fix the formatting

**Comment by mattdangerw** [comment]
Actually, since my comments add up to just a couple lines changes, I can just make these as merge this. Thanks very much for contribution!

**Pull request merged**